service/job/job_chord.py

```diff
@@ -0,0 +1,48 @@
+import uuid
+from celery import chord, signature
+
+from utils import job_db
+
+
+class JobChord:
+    """Wraps multiple celery task chains as a celery chord and handles ID generation."""
+
+    def __init__(self, chains):
+        """
+        Create a job chord and triggers ID generation.
+
+        :param List[chain] chains: A list of celery task chains
+        """
+        self.id = str(uuid.uuid1())
+
+        # Once all job chains have finished within this chord, we have to trigger a
+        # callback worker in order to update the database entry for the chord job itself.
+        self.chord = chord(chains, signature(
+            'finish_chord', kwargs={'job_id': self.id, 'work_path': self.id}))
+
+        self.chordSubtaskIds = []
+
+        for current_chain in self.chord.tasks:
+            current_chain_ids = []
+            current_chain_work_path = str(uuid.uuid1())
+            current_chain.kwargs['work_path'] = current_chain_work_path
+
+            for single_task in current_chain.tasks:
+                task_id = str(uuid.uuid1())
+
+                single_task.kwargs['job_id'] = task_id
+                single_task.kwargs['work_path'] = current_chain_work_path
+                single_task.options['task_id'] = task_id
```

Comment:
Hier setze ich noch die job_id, weil sonst die Worker crashen. Ich hatte gesehen dass Sebastian irgendwann neulich mal ein Refactoring vorgenommen hatte, bei dem er die job_ids von den work_paths getrennt hat. Kann es sein dass man da die Überprüfung der job_ids in den Workern inzwischen raus nehmen könnte? Ich fände es für die Gesamtarchitektur irgendwie eleganter, wenn nur die "top level" Jobs auch eine ID bekommen. Alles was nicht direkt Ergebnis einer Eingabe im Frontend ist (alle anderen Subtasks) bekommen dann nur noch Task IDs. Vielleicht übersehe ich gerade aber auch einen Grund, weswegen wir die job_ids weiter brauchen.

Reply:
Ok, dann glaube ich sind die inzwischen hinfällig. 

Das ursprüngliche Problem war vermutlich:
Chains haben keine eigene Task ID, sondern konstruieren im Grunde nur Pointer zwischen einzelnen Tasks. Um also bei FinishJob oder Error der "verketteten" Subtasks die Chain ansprechen zu können, war es nötig diese mit einem Job ID parameter zu versehen. Kommt das hin?

Die alte Variante war also: Chain (Job ID) -> mehrere Tasks (Job ID + individuelle Task ID).
Die neue Variante ist: Chord (Job ID + Task ID) -> mehrere Chains (aktuell gar keine eigene ID) -> pro Chain mehrere Tasks (Job ID + Task ID).

Könnten wir nicht bei Chord und Subtask einfach nur Task IDs per Hand setzen anstatt Job IDs zu nutzen? Wenn beim Chord ein Subtask failed, failed auch der Chord. Das heißt die Subtasks brauchen gar keinen Hinweis mehr darauf, wie sie Fehler "nach oben" melden. Damit wäre eine separate Job ID neben den Task IDs hinfällig.